src/voter_set.rs

```diff
@@ -92,46 +77,70 @@ impl<Id: Eq + Ord> VoterSet<Id> {
 
 		let total_weight = VoterWeight::new(total_weight).expect("voters nonempty; qed");
 
-		// Establish the total order based on the voter IDs.
-		let order = voters.keys().cloned().collect::<Vec<_>>();
-		for (i, id) in order.iter().enumerate() {
-			voters.get_mut(id).expect("def. of order; qed").position = i;
+		// SAFETY: Order of same id doesn't matter as they will be aggregated.
+		// This is like tim-sort: if pre-sorted will be extremely efficient.
+		voters.sort_unstable_by(|(a_id, _), (b_id, _)| a_id.cmp(b_id));
+
+		// voters contains neighboring duplicates. Reduce list so all weight are against one id
+		let mut position = 0;
+		let mut current_weight = 0;
+		for i in 0..voters.len() {
+			let (id, VoterInfo { weight: VoterWeight(weight), .. }) = &voters[i];
+			current_weight += weight.get();
+			debug_assert!(current_weight != 0);
+
+			// if id doesn't match next_id or is last in list:
+			if voters.get(i + 1).filter(|(next_id, _)| id == next_id).is_none() {
+				voters[position] = (
+					id.clone(),
+					VoterInfo {
+						position,
+						weight: VoterWeight::new(current_weight).expect("non-zero"),
+					},
+				);
+				position += 1;
+				current_weight = 0;
+			}
 		}
 
-		let threshold = threshold(total_weight);
+		// Duplicate ids and weights after this position have already been processed.
+		voters.truncate(position);
 
-		Some(VoterSet { voters, order, total_weight, threshold })
+		Some(VoterSet { voters, total_weight, threshold: threshold(total_weight) })
```

Comment:
Can we keep the original code for aggregating with a `BTreeMap` and then do a simple conversion to `Vec<(Id, VoterInfo)>`? The code will be simpler and the voter set constructor isn't run that often so we don't really need to optimize it.

Reply:
Better?